helios-similar.man
.TH helios-similar 1 "June 2019"
.SH NAME
helios-similar - Search for similar songs on a remote Helios server.

.SH SYNOPSIS
.B helios-similar [\fIOPTIONS\fR] [--file="<file>" | --url="<url>" | --id="<id>" | --reference="<reference>"]

.SH DESCRIPTION
Use this utility to have a remote Helios server perform a similarity match
within its catalogue of analyzed music using a given song as a search key. The
results will be output in JSON format.

The search key is a song which can be either internal to the server's catalogue
or external. If the former you must specify which song to select from the
database with either an --id or a --reference.

If the song is external you must supply either a --file or a --url. The former
submits a song file. The latter submits a song at a given URL hosted on
Bandcamp, Mixcloud, SoundCloud, Vimeo, YouTube, or any of the numerous other
supported external services.

All songs used as external search keys are not stored on the server after their
analysis. They are streamed into the server's temporary storage and then
immediately deleted after analysis.

.SH OPTIONS

.TP
\fB\--file="<file>"\fR
Path to a local song file to use as a search key on the server. This performs
the query using an external search key. You must provide exactly one of --file,
--id, --reference, or --url.

.TP
\fB\--id="<song_id>"\fR
Unique numeric identifier of song already within the database to use as a search
key. You must provide exactly one of --file, --id, --reference, or --url.

.TP
\fB\--results="<size>"\fR
Maximum number of similarity results to return. Default is ten.

.TP
\fB\--reference="<song_reference>"\fR
Unique reference of song already within the database to use as a search key.
You must provide exactly one of --file, --id, --reference, or --url.

.TP
\fB\--short\fR
Display results in short form without any JSON as simply "Artist - Title" format.

.TP
\fB\--url="<url>"\fR
URL of a song hosted on any of a number of supported external services to use as
a search key on the server. These include Bandcamp, Mixcloud, SoundCloud, Vimeo,
and YouTube among many others. You must provide exactly one of --file, --id,
--reference, or --url.

.so common.man.include

.SH EXAMPLES
.TP
$ helios-similar --url "https://emika-official.bandcamp.com/track/battles"

.TP
$ helios-similar --file "Emika - Battles.flac"

.TP
$ helios-similar --id 1773

.SH EXIT STATUS
\fBhelios-similar\fR exits with a status of zero if the server provided the expected response or 1 otherwise.

.SH AUTHOR
Cartesian Theatre <info@cartesiantheatre.com>

.SH REPORTING BUGS
Report bugs to Cartesian Theatre <info@cartesiantheatre.com>

.SH COPYRIGHT
Copyright (C) 2015-2019 Cartesian Theatre. All rights reserved.

.SH SEE ALSO
\fBhelios\fR(7)
.br
\fBhelios-get-song\fR(1)
.br
\fIhttps://www.heliosmusic.io\fR



helios-similar.py
#!/usr/bin/python3
#
#   Helios, intelligent music.
#   Copyright (C) 2015-2019 Cartesian Theatre. All rights reserved.
#

# System imports...
import helios
from helios.responses import StoredSongSchema
from common import add_common_arguments, zeroconf_find_server
from hfilesize import Format, FileSize
from pprint import pprint
from termcolor import colored
import argparse
import attr
import colorama
import datetime
import sys

# i18n...
import gettext
_ = gettext.gettext

# Add arguments specific to this utility to argument parser...
def add_arguments(argument_parser):

    # Add a mutually exclusive group for song search key selection...
    search_key_selection_group = argument_parser.add_mutually_exclusive_group(required=True)

    # Define behaviour for --file in song search key selection exclusion group...
    search_key_selection_group.add_argument(
        '--file',
        dest='similar_file',
        required=False,
        help=_('Path  to a local song file to use as a search key on the server.'))

    # Define behaviour for --id in song search key selection exclusion group...
    search_key_selection_group.add_argument(
        '--id',
        dest='similar_id',
        required=False,
        nargs='?',
        help=_('Unique  numeric  identifier  of song already within the database to use as a search key.'),
        type=int)

    # Define behaviour for --reference in song search key selection exclusion group...
    search_key_selection_group.add_argument(
        '--reference',
        dest='similar_reference',
        required=False,
        nargs='?',
        help=_('Unique reference of song already within the database to use as a search key.'))

    # Define behaviour for --url in song search key selection exclusion group...
    search_key_selection_group.add_argument(
        '--url',
        dest='similar_url',
        required=False,
        nargs='?',
        help=_('URL  of  a  song  hosted  on  any of a number of supported external services to use as a search key on the server.'))

    # Define behaviour for --results...
    argument_parser.add_argument(
        '--results',
        dest='maximum_results',
        default=10,
        required=False,
        nargs='?',
        help=_('Maximum number of similarity results to return. Default is ten.'),
        type=int)

    # Define behaviour for --short...
    argument_parser.add_argument(
        '--short',
        action='store_true',
        default='False',
        dest='short',
        help=_('Display results in short form without any JSON as simply \"Artist - Title\" format.'))


# Entry point...
if __name__ == '__main__':

    # Initialize the argument parser...
    argument_parser = argparse.ArgumentParser(
        description=_('Search for similar songs on a remote Helios server.'))

    # Add common arguments to argument parser...
    add_common_arguments(argument_parser)

    # Add arguments specific to this utility to argument parser...
    add_arguments(argument_parser)

    # Parse the command line...
    arguments = argument_parser.parse_args()

    # Initialize terminal colour...
    colorama.init()

    # Status on whether there were any errors...
    success = False

    # Try to retrieve metadata...
    try:

        # If no host provided, use Zeroconf auto detection...
        if not arguments.host:
            arguments.host = zeroconf_find_server()[0]

        # Create a client...
        client = helios.client(
            host=arguments.host,
            port=arguments.port,
            token=arguments.token,
            verbose=arguments.verbose)

        # Prepare request parameters...
        similarity_search_dict = {}
        if arguments.similar_file:
            similarity_search_dict['similar_file'] = base64.b64encode(open(arguments.similar_file, 'rb').read())
        if arguments.similar_id:
            similarity_search_dict['similar_id'] = arguments.similar_id
        if arguments.similar_reference:
            similarity_search_dict['similar_reference'] = arguments.similar_reference
        if arguments.similar_url:
            similarity_search_dict['similar_url'] = arguments.similar_url
        similarity_search_dict['maximum_results'] = arguments.maximum_results

        # Query...
        similar_songs_list = client.get_similar_songs(similarity_search_dict)

        # Note success...
        success = True

        # Create a schema to deserialize stored song objects into JSON...
        stored_song_schema = StoredSongSchema()

        # Display each song and end with a new line...
        for song in similar_songs_list:

            # If we are not using short form output, display each song in JSON
            #  format...
            if not arguments.short:
                pprint(stored_song_schema.dump(song))
                print('')

            # Otherwise show it in short form...
            else:
                print(F'{song.artist} - {song.title}')

    # User trying to abort...
    except KeyboardInterrupt:
        sys.exit(1)

    # Helios exception...
    except helios.exceptions.ExceptionBase as someException:
        print(someException.what())

    # Some other kind of exception...
    except Exception as someException:
        print(_(f"An unknown exception occurred: {print(someException)}"))

    # If unsuccessful, bail...
    if not success:
        sys.exit(1)

    # Done...
    sys.exit(0)

